### PR TITLE
Refactor and Introduce ExtensionManager in CodeGen

### PR DESCRIPTION
**Pull request by ia3andy**
This is part of #8178

- Improved packages names and content
- Improved general readability and consistency
- Added ExtensionManager to define a high level way of managing (read/write) extensions in any QuarkusProject
- Removed most unsafe Gradle operations which were outside of the Gradle plugin (we need to figure out a way to improve the "generic" gradle support to make it compatible again..)
- Removed existing project support with the create command (throws an error) -> https://github.com/quarkusio/quarkus/issues/9875
- Removed compatibility with project without the Quarkus platform bom defined

### CODE REVIEW

**Comment by ia3andy** [comment]
I need to seperate the moving from the modification to make it easier to review..

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java
Comment (gastaldi) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java:69]: Do they have to be `Extension` or the `AppArtifactKey` is enough? Also, perhaps it would be nice to use `Collection` instead

**Review by aloubyansky — APPROVED** [review]

File: devtools/maven/src/main/java/io/quarkus/maven/CreateProjectMojo.java
Comment (stalep) [inline devtools/maven/src/main/java/io/quarkus/maven/CreateProjectMojo.java:221]: Why remove the ability to add extensions when you create a project?
I find the feature quite useful.

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java
Comment (ia3andy) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java]: Note that I renamed the different methods since your comment (install, getInstalled, uninstall..), I wanted to do it before to push, but I rushed a bit and forgot..

@gastaldi this is a good question, TBH I am not sure, what I think is important is to make sure that the input is well formatted, and not just some random string (that are resolved) like we get it the commands. At the end, for install, we need either the GA or the GAV (if the extension is not managed) and for uninstall just the GA, for `getInstalled` the GAV

Regarding `Set` or `Collection`, it's going to be converted to a Set anyway to remove duplicates..

File: devtools/maven/src/main/java/io/quarkus/maven/CreateProjectMojo.java
Comment (ia3andy) [reply devtools/maven/src/main/java/io/quarkus/maven/CreateProjectMojo.java:221]: @stalep it's done directly from the create command (no need to call two separate commands)

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java
Comment (aloubyansky) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java]: BTW, even if the extension is present in the platform, a user should still be able to override its version.

**Comment by gastaldi** [comment]
This change impacts #9593, I'll wait until this is merged to proceed

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java
Comment (ia3andy) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java]: @aloubyansky I am not 100% sure, but I think we said that since it's not a "recommended" action, we would advise the user to do it manually?

A possible middle group would be to when the specified version is different from the platform one, we add it, but show a warning?

Comment (aloubyansky) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java]: I don't think it costs us anything to respect the version choice.

**Comment by gastaldi** [comment]
@stalep I think it will probably affect the classes formerly known as 'ListExtensionsCommandHandler` and `AddExtensionsCommandHandler` :)

**Comment by ia3andy** [comment]
I am in PTO tomorrow, I will not be able to follow up on the discussions before monday.. really sorry :-/

@stalep @gastaldi can your PRs wait until then?

**Comment by gastaldi** [comment]
@ia3andy sure, I'll be refining the registry descriptor API meanwhile :wink:

**Comment by stalep** [comment]
No problem, I just promised @maxandersen I would do a PR today :)

**Comment by maxandersen** [comment]
@stalep do open the pr - better to have and then we can help rebase and get early feedback started.

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java
Comment (maxandersen) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java]: it is recommended to use a specific version when that is what is needed for the users app.
Thus if user are being explicit we should carry that information through the system to honor it.

File: devtools/gradle/src/main/java/io/quarkus/gradle/GradleBuildFileFromConnector.java
Comment (maxandersen) [inline devtools/gradle/src/main/java/io/quarkus/gradle/GradleBuildFileFromConnector.java:45]: remove if it should be ignored.

File: devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusPlatformTask.java
Comment (maxandersen) [inline devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusPlatformTask.java:70]: this code could be more readable to put `getProject().getRootProject().getProjectDir()`  in var.

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/commands/handlers/QuarkusCommandHandlers.java
Comment (maxandersen) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/commands/handlers/QuarkusCommandHandlers.java:71]: why the change of signature here ?

Comment (ia3andy) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/commands/handlers/QuarkusCommandHandlers.java:71]: TBH I don't remember 😅 the name is clearer, but the order 🤗, I'll switch the order back..

**Comment by ia3andy** [comment]
I made modifications as discussed, I need to fix all the tests, I am going to push later in the afternoon..

File: devtools/gradle/src/main/java/io/quarkus/gradle/GradleBuildFileFromConnector.java
Comment (ia3andy) [reply devtools/gradle/src/main/java/io/quarkus/gradle/GradleBuildFileFromConnector.java:45]: @maxandersen this is actually existing code, because of too much diff, git won't detect that it's the same file.

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java
Comment (ia3andy) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java]: ok user version is now taken into account and I used `AppArtifactKey`

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/commands/handlers/QuarkusCommandHandlers.java
Comment (ia3andy) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/commands/handlers/QuarkusCommandHandlers.java]: fixed

File: independent-projects/tools/platform-descriptor-api/src/main/java/io/quarkus/dependencies/Extension.java
Comment (ia3andy) [inline independent-projects/tools/platform-descriptor-api/src/main/java/io/quarkus/dependencies/Extension.java:199]: @aloubyansky as we discussed ⬆️

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/model/AppArtifactKey.java
Comment (ia3andy) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/model/AppArtifactKey.java:76]: @aloubyansky as we discussed ⬆️

**Comment by ia3andy** [comment]
ok pushed :)

File: devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusPlatformTask.java
Comment (ia3andy) [reply devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusPlatformTask.java]: Yeah that's better indeed!

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/buildfile/MavenBuildFile.java
Comment (ia3andy) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/buildfile/MavenBuildFile.java:49]: @aloubyansky maybe we should change the AppArtifactKey/Coords instead, having empty string classifier seems not working with the maven api?

Comment (aloubyansky) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/buildfile/MavenBuildFile.java:49]: What do you mean? The only reason we use an empty classifier is because this is how Maven does it.

**Comment by ia3andy** [comment]
@maxandersen @aloubyansky @gastaldi this PR is ready to be review, there is just an old IT not passing, it will take another hour before to pass (nothing that would changes the PR fundamentally)

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/buildfile/MavenBuildFile.java
Comment (ia3andy) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/buildfile/MavenBuildFile.java:49]: We discussed on zulip and concluded that maven api is behavior differently for resolving artifact and dependencies. Making sure we don't put an empty classifier in the dependency prevent from having `<classifier></classifier>` in the pom.xml.

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java
Comment (gastaldi) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java:15]: Better use the singular form here: `ExtensionManager`. eg. https://github.com/quarkusio/quarkus-registry-descriptor/blob/master/src/main/java/io/quarkus/registry/ExtensionRegistry.java

Comment (ia3andy) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionsManager.java:15]: ok I'll change

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/commands/handlers/QuarkusCommandHandlers.java
Comment (gastaldi) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/commands/handlers/QuarkusCommandHandlers.java:48]: Since you're refactoring `Printer` into `ConsoleMessageFormat`, try to not expose the `NOK` attribute there otherwise the methods in that class are meaningless

File: independent-projects/tools/platform-descriptor-api/src/main/java/io/quarkus/platform/tools/ConsoleMessageFormat.java
Comment (gastaldi) [inline independent-projects/tools/platform-descriptor-api/src/main/java/io/quarkus/platform/tools/ConsoleMessageFormat.java:7]: Better make these attributes private otherwise any methods in this class are meaningless

File: integration-tests/maven/src/test/java/io/quarkus/maven/it/CreateProjectMojoIT.java
Comment (gastaldi) [inline integration-tests/maven/src/test/java/io/quarkus/maven/it/CreateProjectMojoIT.java:354]: Why was this test removed?

File: integration-tests/maven/src/test/java/io/quarkus/maven/it/CreateProjectMojoIT.java
Comment (gastaldi) [inline integration-tests/maven/src/test/java/io/quarkus/maven/it/CreateProjectMojoIT.java:147]: It looks like those asserts are important, why remove them?

Comment (ia3andy) [reply integration-tests/maven/src/test/java/io/quarkus/maven/it/CreateProjectMojoIT.java:147]: Migrating project has been removed from the "create" command, those asserts were for it. I am now testing that it fails. Note that the same asserts are already in the other "fromScratch" test.

File: integration-tests/maven/src/test/java/io/quarkus/maven/it/CreateProjectMojoIT.java
Comment (ia3andy) [reply integration-tests/maven/src/test/java/io/quarkus/maven/it/CreateProjectMojoIT.java:354]: same as the previous

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java
Comment (gastaldi) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java:63]: I'd prefer to use `Collection` instead. That would prevent cases where the caller has a different Collection implementation and need to convert to List to call this method

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java
Comment (gastaldi) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java:74]: Same as above

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java
Comment (gastaldi) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java:28]:
```suggestion
    Collection<AppArtifactCoords> getInstalled() throws IOException;
```

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java
Comment (ia3andy) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java:63]: For this one I would really like to have a List since it's ordered and you may want to add them in a specific order..

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java
Comment (ia3andy) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java:74]: I already answer for this, the Set already takes care of deduplicating, I would convert it to a `Set` just after if that's was not the case. I think it's ok this way.. @aloubyansky any thoughts?

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java
Comment (gastaldi) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java:99]: Instead of returning how many extensions were installed/uninstalled wouldn't be better to return *which* extensions were installed/uninstalled (the collection size would return the total of installed/uninstalled items)?

Comment (ia3andy) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java:99]: @gastaldi that's a good idea! and it could be done without changing the contract. Let's do it in another PR.

Comment (ia3andy) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java:99]: yeah and I need to print it too, it's not printed anymore when installing/uninstalling.. i'll create an issue as soon as we merge that one.

**Review by gastaldi — CHANGES_REQUESTED** [review]
Looks good, but the PR needs to be squashed before merged

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/commands/AddExtensions.java
Comment (gastaldi) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/commands/AddExtensions.java:23]:
```suggestion
    public static final String EXTENSION_MANAGER = ToolsUtils.dotJoin(ToolsConstants.QUARKUS, NAME, "extension-manager");
```

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/buildfile/BuildFile.java
Comment (gastaldi) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/buildfile/BuildFile.java:156]: Don't expose this method, because the Registry won't be treated as a List of Extensions in the future

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/codegen/writer/FileProjectWriter.java
Comment (gastaldi) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/codegen/writer/FileProjectWriter.java:26]: Nice!

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/codegen/writer/FileProjectWriter.java
Comment (gastaldi) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/codegen/writer/FileProjectWriter.java:31]:
```suggestion
            throw new IOException("Project root needs to be a directory");
```

**Comment by ia3andy** [comment]
@gastaldi I am waiting to see if CI passes then I'll merge your changes (probably tomorrow).. cheers

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java
Comment (stalep) [reply independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/extensions/ExtensionManager.java:74]: The use of Set might be because of making sure the items are unique? - if not I'm also for unifying on either List/Set/Collection.

**Comment by ia3andy** [comment]
@gastaldi changes applied (even the `Collection` 😅) lol and commits squashed

**Review by gastaldi — APPROVED** [review]

**Comment by ia3andy** [comment]
I am trying to figure out why CI is failing (and giving me headaches)

**Review by maxandersen — COMMENTED** [review]
I haven't reviewed this in full depth thus I don't feel right to explictily approve it but @gastaldi and @aloubyansky  seem on top of it; but overall looks as good cleanup and I didn't spot anything alarming.

Just check if `mvn io.quarkus:quarkus-maven-plugin:1.5.0.Final:create --platformGroupId=xxx --platformArtifactid=xxx --platformVersion=xxx` where xxx is pointing to this version of built quarkus platform.


If need deeper review let me know :)

**Comment by ia3andy** [comment]
@gsmet can we merge without JDK14 tests passing?

**Review by aloubyansky — APPROVED** [review]

**Pull request merged**
